### ISSUE TITLE
Separate completion checking from block acquisition

### ISSUE THREAD

**Issue by rhoadesScholar** [funkelab/daisy#52]
In the current implementation, blocks are only checked for completion once they are requested by workers. This checking should instead be run in parallel, such that the master process is eliminating already completed blocks from the queue independently of the worker operations.

https://github.com/funkelab/daisy/blob/d60ef4577c8de3aee61dd5eda2942df8db4b9457/daisy/scheduler.py#L92

**Comment by pattonw** [comment]
~~The scheduler is designed to run in its own thread, so it is already running in parallel~~ The scheduler runs in the main thread with the server. We could just move the `__precheck(block)` to run on release instead of request since the scheduler already keeps a list of available blocks. Essentially changing the order from:
1) recieve request
2) get a block
3) check if block has been completed, if so add newly available blocks to pending list and go back to 2
4) return block

To:
1) recieve request
2) get a block
3) return block
4) check newly available blocks with `__precheck`
5) only append uncompleted newly available blocks to pending blocks

**Comment by pattonw** [comment]
Ah, nevermind. That barely improves things since you would just be waiting for the scheduler after releasing a block, until you can acquire the next block.
I think the best approach would be to just have a worker pool to run the check block function the same way we have a worker pool to run the process block function

**Comment by cmalinmayor** [comment]
## TL;DR Recommendation from Caroline and Will @rhoadesScholar 
1. Speed up your precheck function. This is most important. Ideally the pre-check function could run through all the blocks in seconds. Will has a suggestion involving writing a zarr with a single value in each block when completed (so it will only make one (compressed) file per block).
2. Run pre-check in your process function rather than in the server. If your pre-check must be slow, it should become part of your worker task.

# Notes from discussion:
## Re: running pre-check on the server
 
> This checking should instead be run in parallel, such that the master process is eliminating already completed blocks from the queue independently of the worker operations

There would be two ways to implement this. Both of which are bad (we think). 
     a) Don't spawn more workers. This is what is happening now, essentially. The server or the scheduler have to run `__precheck` on all blocks and the workers wait until this returns a non-completed block 
     b) Have a new pool of workers to run `__precheck`. This speeds up the waiting time problem, but introduces a new one, which is overhead of communication between workers and the server. These pre-check workers clogging the communication would slow down the communication with the actual task workers. Exactly how these trade off depends on the speed of your pre-check function and also how many workers you have.

## Running pre-check on the workers
A solution for long-running pre-check functions is removing it from the master process and running pre-check on the workers before the provided process function (or making people add it to their process function). This is better than b) but could potentially still have issues. It does not double the number of workers talking to the server (yay!). It still causes fast communication when pre-check returns quickly at the beginning of re-processing a dataset, which can slow down the server (although it would still be significantly faster than re-running the actual processing). Note: we vaguely remember issues like dying workers when they are starved of communication and having to be restarted, which could add more overhead. But, we would need to test this.

**Comment by rhoadesScholar** [comment]
Let's assume that any `__precheck` is _very_ quick - some variety of checking a database (otherwise it should be in the process function). 💯% agree that pushing any such function into external workers would unnecessarily clog the TCP communications. However, even for a very quick check function, assuming it must be iterative, if you have enough blocks this will eventually become a bottleneck.

I think @pattonw's proposed reordering would fix the problem. 👍🏼 

> 1. recieve request
> 2. get a block
> 3. return block
> 4. check newly available blocks with `__precheck`
> 5. only append uncompleted newly available blocks to pending blocks

**Comment by pattonw** [comment]
TL;DR:
Caroline's recommendations are also mine.
1) speed up your `__precheck` function.
2) move it into the `process_function` if you can't speed it up.

Working on the reordering and talking to Caroline definitely changed my mind. Reordering just shifts the wait time around. The scheduler does not run in parallel, so you can't return the checked block and then run the `__precheck` function on the newly released blocks. You would have to run the `__precheck` before returning a block, or before releasing a block. Either way a worker is waiting.

Since all blocks are scheduled in a consistent order, rerunning will usually have to iterate through all complete blocks before starting a worker. So starting a single worker to run the scheduler in parallel would have negligible benefit since you would just have to wait for it to check all completed blocks until it finds the first ready block anyway.

So if serial, and single worker parallelism are out, a worker pool is the most reasonable option. Starting a separate pool for `__precheck` will double the amount of tcp messages sent since every block would have to be sent and retrieved from both the "pre_check" pool and "process_pool". I think this leads to more problems with tcp congestion than it is worth. This leaves us with just keeping the current pool and executing the `check_function` in the `process_function`. This is my recommendation for you.

We could move the `__precheck` function into the worker function by default but I don't think this is a good idea. Assuming the `__precheck` is nearly instant (this is the point of `__precheck`, build some lookup table s.t. that it can quickly tell you if a block is done), the tcp communication may be slower than just iterating in the main thread.

**Comment by rhoadesScholar** [comment]
Interesting. It seems like there should be some way to improve performance by decoupling queuing of uncompleted blocks from the actual passing of blocks to the workers, but perhaps not. Fair enough. Thanks!
